zerver/tests/test_message_edit.py

```diff
@@ -1068,33 +1068,271 @@ def test_notify_old_thread_move_message_to_stream(self) -> None:
+        admin_user = self.example_user("iago")
+        user_losing_access = self.example_user('cordelia')
+        user_gaining_access = self.example_user('hamlet')
+
+        self.login("iago")
+        old_stream = self.make_stream("test move stream")
+        new_stream = self.make_stream("new stream", invite_only=True, history_public_to_subscribers=True)
+
+        self.subscribe(admin_user, old_stream.name)
+        self.subscribe(user_losing_access, old_stream.name)
+
+        self.subscribe(admin_user, new_stream.name)
+        self.subscribe(user_gaining_access, new_stream.name)
+
+        msg_id = self.send_stream_message(admin_user, old_stream.name,
+                                          topic_name="test", content="First")
+        self.send_stream_message(admin_user, old_stream.name,
+                                 topic_name="test", content="Second")
+
+        self.assertEqual(UserMessage.objects.filter(
+            user_profile_id=user_losing_access.id,
+            message_id=msg_id,
+        ).count(), 1)
+        self.assertEqual(UserMessage.objects.filter(
+            user_profile_id=user_gaining_access.id,
+            message_id=msg_id,
+        ).count(), 0)
 
+        result = self.client_patch("/json/messages/" + str(msg_id), {
+            'message_id': msg_id,
+            'stream_id': new_stream.id,
+            'propagate_mode': 'change_all',
+        })
+        self.assert_json_success(result)
+
+        messages = get_topic_messages(admin_user, old_stream, "test")
+        self.assertEqual(len(messages), 1)
+        self.assertEqual(messages[0].content, f"This topic was moved by @_**Iago|{admin_user.id}** to #**new stream>test**")
+
+        messages = get_topic_messages(admin_user, new_stream, "test")
+        self.assertEqual(len(messages), 3)
+
+        self.assertEqual(UserMessage.objects.filter(
+            user_profile_id=user_losing_access.id,
+            message_id=msg_id,
+        ).count(), 0)
+        # UserMessage is not created for the user but the user
+        # can see the message since the history is shared.
+        self.assertEqual(UserMessage.objects.filter(
+            user_profile_id=user_gaining_access.id,
+            message_id=msg_id,
+        ).count(), 0)
+
+    def test_move_message_to_stream_private_to_private_stream_not_shared_history(self) -> None:
+        admin_user = self.example_user("iago")
+        user_losing_access = self.example_user('cordelia')
+        user_gaining_access = self.example_user('hamlet')
+
+        self.login("iago")
+        old_stream = self.make_stream("test move stream", invite_only=True)
+        new_stream = self.make_stream("new stream", invite_only=True, history_public_to_subscribers=False)
+
+        self.subscribe(admin_user, old_stream.name)
+        self.subscribe(user_losing_access, old_stream.name)
+
+        self.subscribe(admin_user, new_stream.name)
+        self.subscribe(user_gaining_access, new_stream.name)
```

Comment:
These tests are really long; maybe it's worth writing a helper function that does the common setup and takes a few parameters (e.g. the kwargs to pass to the two `make_stream` calls?)